hpct-ldap-server/src/managers/openldap.py

```diff
@@ -0,0 +1,491 @@
+        dc : str
+            Domain components.
+        """
+        return ",".join([f"dc={x}" for x in dc.split(".")])
+
+    def add_user(
+        self,
+        admin_passwd=None,
+        domain=None,
+        gecos=None,
+        gid=None,
+        homedir=None,
+        shell=None,
+        passwd=None,
+        uid=None,
+        user=None,
+    ) -> None:
+        """Add user.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gecos : str
+            Real name.
+        gid : int
+            Group id.
+        homedir : str
+            Home directory path.
+        shell : str
+            Login shell.
+        passwd : str
+            User password.
+        uid : int
+            User id.
+        user : str
+            Username.
+        """
+        dcs = self._split_domain(domain)
+        if uid < MIN_UID or gid < MIN_GID:
+            logger.error("uid and gid must be below {MIN_UID}.")
+            return
+        if (
+            None not in [admin_passwd, dcs, gecos, gid, homedir, shell, passwd, uid, user]
+        ):
+            binddn = f"cn=admin,{dcs}"
+            base_user = [
+                f"dn: uid={user.lower()},ou=people,{dcs}",
+                "objectClass: inetOrgPerson",
+                "objectClass: posixAccount",
+                "objectClass: shadowAccount",
+                f"cn: {user.lower()}",
+                f"sn: {user}",
+                f"userPassword: {passwd}",
+                f"loginShell: {shell}",
+                f"uidNumber: {uid}",
+                f"gidNumber: {gid}",
+                f"homeDirectory: {homedir}",
+                f"gecos: {gecos}",
+            ]
+            base_user = "\n".join(base_user)
+
+            with tempfile.NamedTemporaryFile(mode="w+t", delete=True) as f:
+                f.write(f"{base_user}")
+                f.flush()
+                cmd = ["ldapadd", "-x", "-D", binddn, "-w", admin_passwd, "-f", f.name]
+
+                rc = subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+                if rc != 0:
+                    raise Exception(
+                        f'Unable to add user. Make sure to run "ldapadd -x -D cn=admin,{dcs} -W -f /etc/ldap/basedn.ldif" first.'
+                    )
+
+    def add_group(
+        self,
+        admin_passwd=None,
+        domain=None,
+        gid=None,
+        group=None,
+    ) -> None:
+        """Add group.
+
+        Parameters
+        ----------
+        admin_passwd : str
+            LDAP password.
+        domain : str
+            Domain name.
+        gid : int
+            Group id.
+        group : str
+            Group name.
+        """
+        dcs = self._split_domain(domain)
+
+        if gid < MIN_GID:
+            logger.error("gid must be below {MIN_GID}.")
+            return
+        if None not in [domain, gid, group, admin_passwd]:
```

Comment:
This could do a check on `None` and early exit.

Reply:
Updated as suggested